File: tape/sdk/typescript/src/chaos/scenarios.ts
// Scenarios — declarative bundles of (faults, invariants, seed).
//
// Two layers of faults:
//   * server  — a named failpoint from the Rust catalogue
//                 (`design-principles/chaos.md §5`); rendered into the
//                 `FAILPOINTS` env-var spec by `failpointsEnv()`.
//   * connector — a wrap around a registered connector in `CONNECTORS`.
//
// The session() context applies connector wraps on enter() and restores
// them on exit(), running invariants against the live journal as it goes.

import { TapeClient } from '../client.ts';
import { CONNECTORS } from '../connectors/index.ts';
import type { Connector } from '../connectors/index.ts';
import { ChaosConnector } from './connectors.ts';

export type FaultLayer = 'server' | 'connector';

export interface Fault {
  readonly layer: FaultLayer;
  readonly target: string;        // failpoint name OR connector name
  readonly action: string;        // panic/sleep/return (server) OR lose_ack/duplicate/delay (connector)
  readonly probability: number;
  readonly afterN: number;        // server only: skip first N hits
  readonly ms: number;            // delay length
  readonly jitter: number;
  readonly when: string;          // free-form selector (Phase 2 CEL)
  readonly extra: Record<string, unknown>;
}

function fault(partial: Partial<Fault> & { layer: FaultLayer; target: string; action: string }): Fault {
  return {
    probability: 1.0, afterN: 0, ms: 0, jitter: 0,
    when: '', extra: {},
    ...partial,
  };
}

// ── server-layer faults ────────────────────────────────────────────────────

export function crash(failpoint: string,
                       opts: { probability?: number; afterN?: number; when?: string } = {}): Fault {
  return fault({ layer: 'server', target: failpoint, action: 'panic',
                  probability: opts.probability ?? 1.0,
                  afterN: opts.afterN ?? 0,
                  when: opts.when ?? '' });
}

export function delay(failpoint: string,
                       opts: { ms: number; probability?: number; jitter?: number; when?: string }): Fault {
  return fault({ layer: 'server', target: failpoint, action: 'sleep',
                  ms: opts.ms, jitter: opts.jitter ?? 0,
                  probability: opts.probability ?? 1.0,
                  when: opts.when ?? '' });
}

export function error(failpoint: string,
                       opts: { msg?: string; probability?: number; when?: string } = {}): Fault {
  return fault({ layer: 'server', target: failpoint, action: 'return',
                  probability: opts.probability ?? 1.0,
                  when: opts.when ?? '',
                  extra: { actionMsg: opts.msg ?? 'chaos' } });
}

// ── connector-layer faults ─────────────────────────────────────────────────

export function loseAck(opts: { connector?: string; tool?: string; probability?: number }): Fault {
  const target = opts.connector ?? opts.tool ?? '';
  if (!target) throw new Error('loseAck: requires connector or tool');
  return fault({ layer: 'connector', target, action: 'lose_ack',
                  probability: opts.probability ?? 0.3,
                  when: opts.tool ? `tool == ${JSON.stringify(opts.tool)}` : '' });
}

export function duplicate(opts: { connector?: string; tool?: string; probability?: number }): Fault {
  const target = opts.connector ?? opts.tool ?? '';
  if (!target) throw new Error('duplicate: requires connector or tool');
  return fault({ layer: 'connector', target, action: 'duplicate',
                  probability: opts.probability ?? 0.05,
                  when: opts.tool ? `tool == ${JSON.stringify(opts.tool)}` : '' });
}

export function delayConnector(opts: { connector: string; ms: number; jitter?: number }): Fault {
  return fault({ layer: 'connector', target: opts.connector, action: 'delay',
                  ms: opts.ms, jitter: opts.jitter ?? 0 });
}

// ── Scenario ───────────────────────────────────────────────────────────────

export interface Scenario {
  readonly name: string;
  readonly faults: readonly Fault[];
  readonly invariants: readonly Invariant[];
  readonly seed: number;
}

export interface Invariant {
  name: string;
  check(opts: { client: TapeClient; runId?: string }): Promise<InvariantResult>;
}

export interface InvariantResult {
  name: string;
  passed: boolean;
  detail: string;
}

export function scenario(opts: {
  name: string;
  faults?: readonly Fault[];
  invariants?: readonly Invariant[];
  seed?: number;
}): Scenario {
  return {
    name: opts.name,
    faults: Object.freeze([...(opts.faults ?? [])]),
    invariants: Object.freeze([...(opts.invariants ?? [])]),
    seed: opts.seed ?? 0,
  };
}

// ── FAILPOINTS env rendering ───────────────────────────────────────────────

function toFailSpec(f: Fault): string {
  let action: string = f.action;
  if (action === 'sleep') action = `sleep(${Math.floor(f.ms)})`;
  else if (action === 'return') action = `return(${String(f.extra.actionMsg ?? 'chaos')})`;
  else if (action === 'print') action = `print(${String(f.extra.actionMsg ?? 'chaos')})`;
  const parts: string[] = [];
  if (f.afterN > 0) parts.push(`${f.afterN}*off`);
  if (f.probability > 0 && f.probability < 1) parts.push(`${f.probability}*${action}`);
  else parts.push(action);
  return `${f.target}=${parts.join('->')}`;
}

/**
 * Render the server-layer faults of `scen` into the `FAILPOINTS` env-var
 * value the chaos-feature tape-server parses at startup. Connector-layer
 * faults are applied in-process via `session()`.
 */
export function failpointsEnv(scen: Scenario): string {
  return scen.faults
    .filter(f => f.layer === 'server')
    .map(toFailSpec)
    .join(';');
}

// ── ChaosReport + Session ──────────────────────────────────────────────────

export interface ChaosReport {
  scenarioName: string;
  seed: number;
  failpointsSpec: string;
  passed: boolean;
  invariantResults: InvariantResult[];
  notes: string[];
}

export interface SessionOpts {
  url?: string;
  runId?: string;
}

// Tiny seeded PRNG (mulberry32). Same seed → same stream — what the
// scenario's `seed` controls.
function makeRng(seed: number): () => number {
  let s = seed >>> 0;
  return () => {
    s = (s + 0x6D2B79F5) >>> 0;
    let t = s;
    t = Math.imul(t ^ (t >>> 15), t | 1);
    t ^= t + Math.imul(t ^ (t >>> 7), t | 61);
    return ((t ^ (t >>> 14)) >>> 0) / 4294967296;
  };
}

export class Session {
  readonly scenario: Scenario;
  readonly url: string;
  runId?: string;
  readonly failpointsSpec: string;
  readonly report: ChaosReport;
  readonly rng: () => number;
  private restores: Array<() => void> = [];

  constructor(scen: Scenario, opts: SessionOpts = {}) {
    this.scenario = scen;
    this.url = opts.url ?? 'tape://localhost:7878';
    this.runId = opts.runId;
    this.failpointsSpec = failpointsEnv(scen);
    this.rng = scen.seed ? makeRng(scen.seed) : () => Math.random();
    this.report = {
      scenarioName: scen.name, seed: scen.seed,
      failpointsSpec: this.failpointsSpec,
      passed: true, invariantResults: [], notes: [],
    };
  }

  setRunId(rid: string): this { this.runId = rid; return this; }

  /** Apply connector wraps. Call before driving the agent. */
  async enter(): Promise<void> {
    const byTarget = new Map<string, Fault[]>();
    for (const f of this.scenario.faults) {
      if (f.layer !== 'connector') continue;
      const arr = byTarget.get(f.target) ?? [];
      arr.push(f);
      byTarget.set(f.target, arr);
    }
    for (const [name, faults] of byTarget) {
      if (!CONNECTORS.has(name)) {
        this.report.notes.push(`connector fault for ${JSON.stringify(name)} skipped: not registered`);
        continue;
      }
      const real = CONNECTORS.get(name);
      const wrapped = new ChaosConnector(real, faults, this.rng);
      CONNECTORS.replace(name, wrapped);
      this.restores.push(() => CONNECTORS.replace(name, real));
    }
  }

  /** Restore connectors and check invariants against the journal. */
  async exit(thrown?: unknown): Promise<void> {
    for (const fn of [...this.restores].reverse()) {
      try { fn(); } catch { /* swallow */ }
    }
    this.restores = [];
    if (thrown) {
      this.report.passed = false;
      this.report.notes.push(`body raised: ${thrown instanceof Error ? thrown.message : String(thrown)}`);
    }
    const client = new TapeClient(this.url);
    try {
      for (const inv of this.scenario.invariants) {
        let result: InvariantResult;
        try {
          result = await inv.check({ client, runId: this.runId });
        } catch (ex) {
          result = { name: inv.name, passed: false,
                     detail: `check threw: ${ex instanceof Error ? ex.message : String(ex)}` };
        }
        this.report.invariantResults.push(result);
        if (!result.passed) this.report.passed = false;
      }
    } finally {
      client.close();
    }
  }
}

export function session(scen: Scenario, opts: SessionOpts = {}): Session {
  return new Session(scen, opts);
}

/** Run `body(sess)` under `scen`, returning the report. */
export async function runScenario(
  scen: Scenario,
  body: (sess: Session) => Promise<void> | void,
  opts: SessionOpts = {},
): Promise<ChaosReport> {
  const sess = new Session(scen, opts);
  await sess.enter();
  let thrown: unknown;
  try { await body(sess); } catch (e) { thrown = e; }
  await sess.exit(thrown);
  return sess.report;
}

File: tape/sdk/go/chaos/scenarios.go
// Package chaos — TapeChaos for Go. Mirrors `tape.chaos` from the
// Python SDK and `tape-ts/chaos` from the TS SDK. The shape: scenarios
// are declarative bundles of (faults, invariants, seed); faults target
// either the server's failpoint catalogue (Phase 0) or a registered
// connector; invariants are predicates over Tape's journal projections.
//
// See `design-principles/chaos.md` for the full design.
//
// Headline pattern:
//
//	import "github.com/vamsiramakrishnan/durable-agents/tape/sdk/go/chaos"
//
//	scen := chaos.Scenario{
//	    Name: "bank-wire-survives-crash",
//	    Seed: 42,
//	    Faults: []chaos.Fault{
//	        chaos.Crash("tape::begin_effect::post_db", chaos.WithAfterN(1)),
//	        chaos.LoseAck("bank.wire", 0.3),
//	    },
//	    Invariants: []chaos.Invariant{chaos.NoStuckObligations},
//	}
//	sess := chaos.NewSession(scen, chaos.SessionOpts{URL: "tape://127.0.0.1:7878"})
//	if err := sess.Enter(ctx); err != nil { ... }
//	defer sess.Exit(ctx, nil)
//	if err := runMyAgent(ctx, sess); err != nil { sess.Exit(ctx, err); ... }
//	fmt.Println(sess.Report)
package chaos

import (
	"context"
	"fmt"
	"math/rand/v2"
	"strings"
	"sync"

	"github.com/vamsiramakrishnan/durable-agents/tape/sdk/go/connectors"
	tape "github.com/vamsiramakrishnan/durable-agents/tape/sdk/go"
)

// FaultLayer — the layer a Fault targets.
type FaultLayer string

const (
	LayerServer    FaultLayer = "server"
	LayerConnector FaultLayer = "connector"
)

// Fault — one declared chaos rule. Two layers: server (a named failpoint
// in the Rust catalogue) and connector (a wrap around a registered
// connector).
type Fault struct {
	Layer       FaultLayer
	Target      string // failpoint name OR connector name
	Action      string // panic/sleep/return (server) OR lose_ack/duplicate/delay (connector)
	Probability float64
	AfterN      int     // server only: skip first N hits
	Ms          int     // delay length
	Jitter      float64 // ± fraction
	When        string  // selector (Phase 2 CEL); recorded for the report
	ActionMsg   string  // for `error(msg=...)`
}

// FaultOpt — option for fault constructors.
type FaultOpt func(*Fault)

// WithProbability — set the per-hit firing probability (default 1.0).
func WithProbability(p float64) FaultOpt { return func(f *Fault) { f.Probability = p } }

// WithAfterN — fail-crate's `n*off->...` idiom: skip the first n hits.
func WithAfterN(n int) FaultOpt { return func(f *Fault) { f.AfterN = n } }

// WithJitter — ± fraction of the delay (only meaningful for delay/Delay).
func WithJitter(j float64) FaultOpt { return func(f *Fault) { f.Jitter = j } }

// WithWhen — free-form selector recorded for the report (Phase 2 CEL).
func WithWhen(w string) FaultOpt { return func(f *Fault) { f.When = w } }

// WithMsg — message for error(...) faults; ignored elsewhere.
func WithMsg(m string) FaultOpt { return func(f *Fault) { f.ActionMsg = m } }

// ── server-layer constructors ────────────────────────────────────────────

// Crash — a server failpoint configured to panic.
func Crash(failpoint string, opts ...FaultOpt) Fault {
	f := Fault{Layer: LayerServer, Target: failpoint, Action: "panic", Probability: 1.0}
	for _, o := range opts {
		o(&f)
	}
	return f
}

// Delay — a server failpoint configured to sleep(ms).
func Delay(failpoint string, ms int, opts ...FaultOpt) Fault {
	f := Fault{Layer: LayerServer, Target: failpoint, Action: "sleep", Probability: 1.0, Ms: ms}
	for _, o := range opts {
		o(&f)
	}
	return f
}

// Error — a server failpoint configured to return(msg).
func Error(failpoint string, opts ...FaultOpt) Fault {
	f := Fault{Layer: LayerServer, Target: failpoint, Action: "return", Probability: 1.0, ActionMsg: "chaos"}
	for _, o := range opts {
		o(&f)
	}
	return f
}

// ── connector-layer constructors ─────────────────────────────────────────

// LoseAck — the connector's Dispatch returns confirmed, but the wrapper
// flips it to unknown so the reconciler resolves via Observe.
func LoseAck(connector string, probability float64) Fault {
	return Fault{Layer: LayerConnector, Target: connector, Action: "lose_ack", Probability: probability}
}

// Duplicate — the connector's Observe returns duplicate instead of confirmed.
func Duplicate(connector string, probability float64) Fault {
	return Fault{Layer: LayerConnector, Target: connector, Action: "duplicate", Probability: probability}
}

// DelayConnector — sleep `ms ± jitter` before the connector's Dispatch.
func DelayConnector(connector string, ms int, opts ...FaultOpt) Fault {
	f := Fault{Layer: LayerConnector, Target: connector, Action: "delay", Probability: 1.0, Ms: ms}
	for _, o := range opts {
		o(&f)
	}
	return f
}

// ── Scenario + types ─────────────────────────────────────────────────────

// Scenario — a named bundle of faults + invariants + seed.
type Scenario struct {
	Name       string
	Faults     []Fault
	Invariants []Invariant
	Seed       int64
}

// InvariantResult — one invariant's outcome.
type InvariantResult struct {
	Name   string
	Passed bool
	Detail string
}

func (r InvariantResult) String() string {
	mark := "OK "
	if !r.Passed {
		mark = "FAIL"
	}
	if r.Detail != "" {
		return fmt.Sprintf("[%s] %s: %s", mark, r.Name, r.Detail)
	}
	return fmt.Sprintf("[%s] %s", mark, r.Name)
}

// Invariant — predicate over Tape's journal projections. The journal is
// the oracle.
type Invariant interface {
	Name() string
	Check(ctx context.Context, client *tape.Client, runID string) InvariantResult
}

// ── FAILPOINTS env rendering ─────────────────────────────────────────────

func toFailSpec(f Fault) string {
	action := f.Action
	switch action {
	case "sleep":
		action = fmt.Sprintf("sleep(%d)", f.Ms)
	case "return":
		msg := f.ActionMsg
		if msg == "" {
			msg = "chaos"
		}
		action = fmt.Sprintf("return(%s)", msg)
	case "print":
		msg := f.ActionMsg
		if msg == "" {
			msg = "chaos"
		}
		action = fmt.Sprintf("print(%s)", msg)
	}
	var parts []string
	if f.AfterN > 0 {
		parts = append(parts, fmt.Sprintf("%d*off", f.AfterN))
	}
	if f.Probability > 0 && f.Probability < 1.0 {
		parts = append(parts, fmt.Sprintf("%g*%s", f.Probability, action))
	} else {
		parts = append(parts, action)
	}
	return fmt.Sprintf("%s=%s", f.Target, strings.Join(parts, "->"))
}

// FailpointsEnv — render the server-layer faults to the FAILPOINTS env-var
// spec the chaos-feature tape-server parses at startup. Connector-layer
// faults are applied in-process via Session.Enter.
func FailpointsEnv(scen Scenario) string {
	var specs []string
	for _, f := range scen.Faults {
		if f.Layer == LayerServer {
			specs = append(specs, toFailSpec(f))
		}
	}
	return strings.Join(specs, ";")
}

// ── ChaosReport + Session ────────────────────────────────────────────────

// ChaosReport — the outcome of one scenario run.
type ChaosReport struct {
	ScenarioName     string
	Seed             int64
	FailpointsSpec   string
	Passed           bool
	InvariantResults []InvariantResult
	Notes            []string
}

func (r ChaosReport) String() string {
	verdict := "PASS"
	if !r.Passed {
		verdict = "FAIL"
	}
	var b strings.Builder
	fmt.Fprintf(&b, "ChaosReport(%q: %s, seed=%d)", r.ScenarioName, verdict, r.Seed)
	for _, ir := range r.InvariantResults {
		fmt.Fprintf(&b, "\n  - %s", ir)
	}
	for _, n := range r.Notes {
		fmt.Fprintf(&b, "\n  ! %s", n)
	}
	return b.String()
}

// SessionOpts — Session configuration.
type SessionOpts struct {
	URL   string
	RunID string
	// Registry — the connector registry the session will wrap. Defaults
	// to `connectors.Default` if nil.
	Registry *connectors.Registry
}

// Session — the context: applies connector wraps on Enter, restores them
// on Exit, and runs invariants on Exit against the live journal.
type Session struct {
	Scenario       Scenario
	URL            string
	RunID          string
	FailpointsSpec string
	Report         ChaosReport
	rng            *rand.Rand
	registry       *connectors.Registry
	restores       []func()
	mu             sync.Mutex
}

// NewSession — construct (do not Enter).
func NewSession(scen Scenario, opts SessionOpts) *Session {
	url := opts.URL
	if url == "" {
		url = "tape://localhost:7878"
	}
	reg := opts.Registry
	if reg == nil {
		reg = connectors.Default
	}
	var r *rand.Rand
	if scen.Seed != 0 {
		r = rand.New(rand.NewPCG(uint64(scen.Seed), uint64(scen.Seed)^0xdead))
	} else {
		r = rand.New(rand.NewPCG(rand.Uint64(), rand.Uint64()))
	}
	return &Session{
		Scenario:       scen,
		URL:            url,
		RunID:          opts.RunID,
		FailpointsSpec: FailpointsEnv(scen),
		Report: ChaosReport{
			ScenarioName:   scen.Name,
			Seed:           scen.Seed,
			FailpointsSpec: FailpointsEnv(scen),
			Passed:         true,
		},
		rng:      r,
		registry: reg,
	}
}

// SetRunID — late-bind the run identifier (for per-run invariants).
func (s *Session) SetRunID(id string) *Session {
	s.RunID = id
	return s
}

// Enter — apply connector wraps. Call before driving the agent.
func (s *Session) Enter(ctx context.Context) error {
	_ = ctx
	s.mu.Lock()
	defer s.mu.Unlock()

	byTarget := map[string][]Fault{}
	for _, f := range s.Scenario.Faults {
		if f.Layer == LayerConnector {
			byTarget[f.Target] = append(byTarget[f.Target], f)
		}
	}
	for name, faults := range byTarget {
		real, err := s.registry.Get(name)
		if err != nil {
			s.Report.Notes = append(s.Report.Notes,
				fmt.Sprintf("connector fault for %q skipped: not registered", name))
			continue
		}
		wrapped := WrapConnector(real, faults, s.rng)
		s.registry.Replace(name, wrapped)
		realCopy := real
		nameCopy := name
		s.restores = append(s.restores, func() {
			s.registry.Replace(nameCopy, realCopy)
		})
	}
	return nil
}

// Exit — restore connectors and check invariants. Pass any error the
// scenario body raised; it's recorded in the report and the report's
// `Passed` flag goes false.
func (s *Session) Exit(ctx context.Context, thrown error) {
	s.mu.Lock()
	for i := len(s.restores) - 1; i >= 0; i-- {
		func() {
			defer func() { _ = recover() }()
			s.restores[i]()
		}()
	}
	s.restores = nil
	s.mu.Unlock()

	if thrown != nil {
		s.Report.Passed = false
		s.Report.Notes = append(s.Report.Notes, fmt.Sprintf("body raised: %v", thrown))
	}

	client, err := tape.Dial(s.URL)
	if err != nil {
		s.Report.Notes = append(s.Report.Notes, fmt.Sprintf("invariant client connect: %v", err))
		return
	}
	defer client.Close()

	for _, inv := range s.Scenario.Invariants {
		ir := safeCheck(ctx, inv, client, s.RunID)
		s.Report.InvariantResults = append(s.Report.InvariantResults, ir)
		if !ir.Passed {
			s.Report.Passed = false
		}
	}
}

func safeCheck(ctx context.Context, inv Invariant, client *tape.Client, runID string) (out InvariantResult) {
	defer func() {
		if r := recover(); r != nil {
			out = InvariantResult{Name: inv.Name(), Passed: false,
				Detail: fmt.Sprintf("check panicked: %v", r)}
		}
	}()
	return inv.Check(ctx, client, runID)
}

// RunScenario — drive `body(ctx, sess)` under `scen` and return the
// report. Convenience wrapper for the most common shape.
func RunScenario(ctx context.Context, scen Scenario,
	body func(context.Context, *Session) error, opts SessionOpts,
) ChaosReport {
	sess := NewSession(scen, opts)
	if err := sess.Enter(ctx); err != nil {
		sess.Report.Notes = append(sess.Report.Notes, fmt.Sprintf("enter: %v", err))
		return sess.Report
	}
	err := body(ctx, sess)
	sess.Exit(ctx, err)
	return sess.Report
}
